From: =?utf-8?b?Ik1hcmNvIFRyZXZpc2FuIChUcmV2acOxbyki?= <mail@3v1n0.net>
Date: Mon, 25 Aug 2025 14:35:57 +0200
Subject: gdm/authPrompt: Preserve the text when switching entry visibility

The text visibility of an auth prompt entry may change dynamically at
runtime, in particular when unlocking the screen with the blank screen
or with the screen shield visible this is happening:
 - The user starts typing the PAM credential (may be secret or not)
 - Under the hood the PAM conversation starts and actually asks user
   for input
 - The entry text is not cleared because we use what the user has typed
   so far as the preemptive secret.
 - The entry is adjusted to follow the requested PAM text visibility
   settings.
   -> This may lead to switching the actual text entry widget in use

Now, if the first entry that PAM requires is "ECHO ON", like it may
happen in the case in which a visible value is requested, we end up
"clearing" what the user as typed instead of preserving the value as
pre-filled value.

Causing the user typing something wrong instead.

Origin: https://gitlab.gnome.org/GNOME/gnome-shell/-/merge_requests/3850
---
 js/gdm/authPrompt.js | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/js/gdm/authPrompt.js b/js/gdm/authPrompt.js
index 251b180..8fc8f98 100644
--- a/js/gdm/authPrompt.js
+++ b/js/gdm/authPrompt.js
@@ -329,10 +329,12 @@ export const AuthPrompt = GObject.registerClass({
             this._mainBox.replace_child(this._entry, this._passwordEntry);
             this._entry = this._passwordEntry;
             this._inactiveEntry = this._textEntry;
+            this._entry.text = this._inactiveEntry.text;
         } else if (!secret && this._entry !== this._textEntry) {
             this._mainBox.replace_child(this._entry, this._textEntry);
             this._entry = this._textEntry;
             this._inactiveEntry = this._passwordEntry;
+            this._entry.text = this._inactiveEntry.text;
         }
         this._capsLockWarningLabel.visible = secret;
 
